### ISSUE TITLE
Support and preferred conventions for hparams vs **kwargs going forward?

### ISSUE THREAD

**Issue by rgmyr** [PyTorchLightning/pytorch-lightning#2385]
#### What is your question?

I'm a little confused about what is currently supported w/r/t `LightningModule.__init__(hparams)` vs `__init__(**kwargs)`, how exactly these fit together, and the direction this project is going w/r/t preferred conventions.

It seems like we're moving towards `**kwargs` so that model params can be more clearly documented. I agree with that sentiment generally, but at the moment I seem to run into various issues either way.

I'd like to understand this so I can write my code such that it is in line with foreseeable future releases, and so that I'll have more confidence submitting PRs when I run into fixable problems.

#### Code

I'm used to using `__init__(hparams); self.hparams = hparams`, but this seems to break `.load_from_checkpoint()` unless I add a line like this:
```
self.hparams = Namespace(**hparams) if type(hparams) is dict else hparams
```
I also ran into problems with `auto_scale_batch_size`, and had to add this:
```
self.batch_size = self.hparams.batch_size
```
Because of  [this line](https://github.com/PyTorchLightning/pytorch-lightning/blob/2411c3be7055c49bc549a796f95c52530c8ce104/pytorch_lightning/trainer/training_tricks.py#L138) in `training_tricks.py`:
```
if not hasattr(model, batch_arg_name):
    raise MisconfigurationException(f'Field {batch_arg_name} not found in `model.hparams`')
```
If I use `**kwargs` instead, I'm not sure where to add them to the argparser in training scripts or `add_model_specific_args`, when you can load models without a `hparams.yaml`, and where they should get saved for later inspection during tuning, etc.

#### What have you tried?

I've read through the docs and poked around through the code base trying to figure some of this out -- still not sure.

#### What's your environment?

 - OS: Linux
 - Packaging: pip install, conda env
 - Version: 0.8.1

**Comment by stas00** [comment]
huggingface/transformers text-classification PL example also run into this problem: https://github.com/huggingface/transformers/pull/6027#discussion_r460431866

**Comment by Borda** [comment]
@nateraw mind have a look at it?

**Comment by nateraw** [comment]
My _personal_ recommendation is to define your init args explicitly + add type hints. Then use argparser to parse the args and pass those in explicitly as well. **Everything is easier if you can explicitly define it**. Thats the bottom line...using kwargs or hparams w/o stating the args _is possible_, but very hard to follow in my opinion.

However, if you need to parameterize different combinations of DataModule + Model, then things get much more tricky. That's why I submitted a PR for a new object called `LightningArgumentParser` that will handle some confusion around that scenario. Please let me know what you think 😄 .
